ipc_service: open-amp: Align VRINGs

carlocaione opened the pull request:
This patchset is doing three things:

1. It is fixing the bogus algorithm to find the optimal number of descriptors for a given memory size.

2. It is changing values for VDEV_STATUS_SIZE and IPC_SERVICE_STATIC_VRINGS_ALIGNMENT to better align to a usual cache line size.

3. RX/TX VRINGs are now correctly aligned to VRING_ALIGNMENT (and cache line alignment).

carlocaione committed:
ipc_service: open-amp: Align VRINGs

This patchset is doing three things:

1. It is fixing the bogus algorithm to find the optimal number of
   descriptors for a given memory size.

2. It is changing values for VDEV_STATUS_SIZE and
   IPC_SERVICE_STATIC_VRINGS_ALIGNMENT to better align to a usual cache
   line size.

3. RX/TX VRINGs are now correctly aligned to MEM_ALIGNMENT (and cache
   line alignment).

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -250,10 +250,34 @@ static int vr_shm_configure(struct ipc_static_vrings *vr, const struct backend_c
+	 *		   +--------------+  vr->shm_addr
+	 *		   |              |
+	 *		   |              |
+	 *		   |   RX BUFS    |
```
I thought RX_BUFS boundaries should be cache-line-aligned as well.

Looking through the code it looks like an imprecise diagram. Buffers seem to be aligned.

carlocaione replied:
Buffers size is defined using DT so the user must take care of that really (I tried to suggest that by adding a comment to the DT property).

hubertmis commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h:
```diff
@@ -130,7 +130,10 @@
+/*
+ * Size of the status region (possibly a multiple of the cache line size).
+ */
+#define VDEV_STATUS_SIZE	(32)
```
Would it be possible to create some static asserts to make sure `VDEV_STATUS_SIZE`, `CONFIG_IPC_SERVIC_STATIC_VRINGS_ALIGNMENT`, and the `zephyr,buffer_size` dt property are aligned to cache lines?

carlocaione replied:
that could be done (maybe in a following PR?).

carlocaione replied:
@hubertmis I actually added a commit to this PR.

carlocaione committed:
ipc_service: open-amp: Add __ASSERTs on cache alignment

Add some asserts to warn the user about unaligned VDEV status area,
VRINGs and buffers.

Signed-off-by: Carlo Caione <ccaione@baylibre.com>

carlocaione commented:
@arnopo gentle ping

arnopo commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h:
```diff
@@ -145,19 +148,20 @@ static inline size_t vq_ring_size(unsigned int num, unsigned int buf_size)
-	num_desc = (unsigned int) (available / single_alloc);
+	while ((available - shm_size(num_desc, buf_size)) >= shm_size(1, buf_size)) {
+		num_desc++;
+	}
```
that's probably true without alignment, with alignment not sure.

carlocaione replied:
Uhm, why not?

`shm_size` is basically returning `VRING_COUNT * (aligned TX/RX buffers + aligned VRINGs)`

carlocaione replied:
I'm still not fully convinced by your remark but your suggestion is pretty much equivalent, so I'll implement that.

arnopo commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -250,10 +250,34 @@ static int vr_shm_configure(struct ipc_static_vrings *vr, const struct backend_c
+	 *		   |   RX BUFS    |
+	 *		   |              |
+	 *		   |              |
+	 *		   +--------------+
```
here would probably important to have also an alignment to avoid buffer corruption.

arnopo reviewed (commented):
@carlocaione , 
Thanks for the headache :)
As usual I have probably over commented the PR  but I'm wondering about the alignment with the  cache.
As both size access to this memory, aligning on cache line seems important.
If no cache addresses should probably be aligned on 32 or 64 bits.
Please find few comments in this way.

Additional request could you rename vq_ring_size() in buff_pool_size() or something similar, for readability?

carlocaione commented on dts/bindings/ipc/zephyr,ipc-openamp-static-vrings.yaml:
```diff
@@ -48,4 +48,5 @@ properties:
       The size of the buffer used to send data between host and remote. Default
-      value is RPMSG_BUFFER_SIZE. This property must be the same for host and remote.
+      value is RPMSG_BUFFER_SIZE. This property must be the same for host and
+      remote and possibly a multiple of the cache line size.
```
hahah ok, I'm not exactly sure about the difference but I trust your suggestion :D

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -250,10 +250,34 @@ static int vr_shm_configure(struct ipc_static_vrings *vr, const struct backend_c
+	 *		   +--------------+
+	 */
+
 	vr->shm_addr = conf->shm_addr + VDEV_STATUS_SIZE;
```
> Seems to me that we should have this aligned on cache line `ROUND_UP(conf->shm_addr + VDEV_STATUS_SIZE, VRING_ALIGNMENT);

Yeah, I kinda assumed that both `shm_addr` and `VDEV_STATUS_SIZE` were already aligned but it doesn't harm to add that.

See: https://github.com/zephyrproject-rtos/zephyr/blob/d8661f151d4b95ccbbd91f9163351688a7ce1b4a/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L764-L765

> And perhaps VRING_ALIGNMENT should be renamed MEMORY_ALIGNMENT ( with a comment specifying that it should be equak to a cache Line or at leat to the core architecture ( Should support alignment on 64-bits)

Good idea.

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -250,10 +250,34 @@ static int vr_shm_configure(struct ipc_static_vrings *vr, const struct backend_c
-	vr->rx_addr = vr->shm_addr + VRING_COUNT * vq_ring_size(num_desc, conf->buffer_size);
+	vr->rx_addr = ROUND_UP(vr->shm_addr +
+			       VRING_COUNT * vq_ring_size(num_desc, conf->buffer_size),
+			       VRING_ALIGNMENT);
```
The point here is that `vq_ring_size` should be aligned because `conf->buffer_size` should be already aligned. The buffer size is set into the DT and while we recommend in the DT binding for the alignment, we also check that at init time in: https://github.com/zephyrproject-rtos/zephyr/blob/d8661f151d4b95ccbbd91f9163351688a7ce1b4a/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L768-L769

carlocaione commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h:
```diff
@@ -145,19 +148,20 @@ static inline size_t vq_ring_size(unsigned int num, unsigned int buf_size)
-	return (VDEV_STATUS_SIZE + (VRING_COUNT * vq_ring_size(num, buf_size)) +
-	       (VRING_COUNT * vring_size(num, VRING_ALIGNMENT)));
+	return (VRING_COUNT * (vq_ring_size(num, buf_size) +
+		ROUND_UP(vring_size(num, VRING_ALIGNMENT), VRING_ALIGNMENT)));
```
Right, but as I wrote in the previous comment the RX / TX buffers should be aligned already, see: https://github.com/zephyrproject-rtos/zephyr/blob/d8661f151d4b95ccbbd91f9163351688a7ce1b4a/subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c#L768-L769

arnopo commented on include/zephyr/ipc/ipc_static_vrings.h:
```diff
@@ -25,8 +25,8 @@ extern "C" {
 
-/** VRING alignment. */
-#define VRING_ALIGNMENT	CONFIG_IPC_SERVICE_STATIC_VRINGS_ALIGNMENT
+/** Memory alignment. This should be aligned to the cache line. */
```
Proposal:
If the cache is enable must be aligned on cache line, else should be aligned on processor 
architecture ( e.g 32-bits).

carlocaione replied:
I'll fix the comment.

arnopo reviewed (dismissed):
LGTM

Thalley commented:
> Putting this in DNM until approved by @Thalley.

```
uart:~$ bt init
Bluetooth init failed (err -11)
[00:00:05.833,740] <inf> fs_nvs: 6 Sectors of 4096 bytes
[00:00:05.833,770] <inf> fs_nvs: alloc wra: 0, fe8
[00:00:05.833,770] <inf> fs_nvs: data wra: 0, 0
[00:00:06.834,838] <err> bt_hci_driver: bt_rpmsg_open: Endpoint binding failed with -11
[00:00:06.834,838] <err> bt_hci_core: bt_enable: HCI driver open failed (-11)
uart:~$ 
```

Applying this patch does break compatibility with one of our 3rd party controllers builds for the nRF5340 netcore.

carlocaione commented:
> Applying this patch does break compatibility with one of our 3rd party controllers builds for the nRF5340 netcore.

Is the 3rd party controller using Zephyr as well?

@hubertmis @arnopo maybe we should hide these changes behind `CONFIG_OPENAMP_WITH_DCACHE` ?

arnopo commented:
> > Applying this patch does break compatibility with one of our 3rd party controllers builds for the nRF5340 netcore.
> 
> Is the 3rd party controller using Zephyr as well?
> 
> @hubertmis @arnopo maybe we should hide these changes behind `CONFIG_OPENAMP_WITH_DCACHE` ?

Would be nice to first understand the issue. No evidence regarding the log (at least for me).

carlocaione commented:
> Would be nice to first understand the issue.

IIUC the issue (@Thalley @hubertmis  please correct me if I'm wrong) is that on the hardware managed by Nordic, only the master can be re-compiled to embed these changes, the slave part of the communication is left untouched (because provided by a 3rd party? Not sure why that part cannot be re-compiled as well).

So, since we are slightly changing the contract between master and slave with this PR when it comes to the memory organization in the SHMEM (because of the alignments), this is probably breaking the communication.

arnopo commented on include/zephyr/ipc/ipc_static_vrings.h:
```diff
@@ -25,8 +25,13 @@ extern "C" {
+ * This should take into account the cache line if the cache is enabled, otherwise
+ * it should be naturally aligned to the machine word size.
+ */
+#define MEM_ALIGNMENT	CONFIG_IPC_SERVICE_STATIC_VRINGS_ALIGNMENT
```
s/CONFIG_IPC_SERVICE_STATIC_VRINGS_ALIGNMENT/CONFIG_IPC_SERVICE_MEM_ALIGNMENT/

carlocaione replied:
I'm going to use `IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT` because this is related to the static VRINGs backend.

arnopo commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.h:
```diff
@@ -130,7 +130,10 @@
+/*
+ * Size of the status region (possibly a multiple of the cache line size).
+ */
+#define VDEV_STATUS_SIZE	(32)
```
s /32/CONFIG_IPC_SERVICE_MEM_ALIGNMENT/

arnopo commented:
To sum-up the change breaks compatibility with Legac... and we can not ensure that  remote core run a Zephyr.

Is the issue is still present if VDEV_STATUS_SIZE CONFIG_IPC_SERVICE_STATIC_VRINGS_ALIGNMENT are set to 4?

carlocaione commented:
> In any case, please do not let our 3rd party use case stop this from being merged, but I thank you for notifying us in time, so that we can prepare for when this is moved to NCS

@Thalley so it is ok with you if we move forward with this?

carlocaione commented:
> But it would still be nice to have some of switch between the two versions, so that we can use and test different versions more easily.

@Thalley  Can you try this PR setting `CONFIG_IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT == 4` ?

Thalley commented:
> > But it would still be nice to have some of switch between the two versions, so that we can use and test different versions more easily.
> 
> @Thalley Can you try this PR setting `CONFIG_IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT == 4` ?

Just did a trial, and unfortunately it still fails with
```
uart:~$ bt init
Bluetooth init failed (err -11)
[00:00:03.012,359] <inf> fs_nvs: 6 Sectors of 4096 bytes
[00:00:03.012,390] <inf> fs_nvs: alloc wra: 0, fe8
[00:00:03.012,420] <inf> fs_nvs: data wra: 0, 0
uart:~$ 
[00:00:04.013,458] <err> bt_hci_driver: bt_rpmsg_open: Endpoint binding failed with -11
[00:00:04.013,458] <err> bt_hci_core: bt_enable: HCI driver open failed (-11)
uart:~$ 
```

carlocaione commented:
@Thalley sorry to be pedantic, could you double check?

I used the nR5340 and the `ipc_sample` with

- NET running Zephyr `main`
- APP running this PR with `CONFIG_IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT == 4`

And they are able to communicate with each other.

More details on the memory regions allocation:

- Using this patch:
```
# VRING0
vr->vring_size: 16
vr->status_reg_addr: 0x20070000
vr->shm_addr: 0x20070020
vr->shm_size: 17344                           
vr->rx_addr: 0x20074020
vr->tx_addr: 0x20074200

#VRING1
vr->vring_size: 64                   
vr->status_reg_addr: 0x20078000
vr->shm_addr: 0x20078020
vr->shm_size: 19840
vr->rx_addr: 0x2007c020
vr->tx_addr: 0x2007c6e0
```

- Using this patch with `CONFIG_IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT == 4`
```
#VRING0
vr->vring_size: 16 
vr->status_reg_addr: 0x20070000
vr->shm_addr: 0x20070004
vr->shm_size: 17248
vr->rx_addr: 0x20074004
vr->tx_addr: 0x200741b4

#VRING1
vr->vring_size: 64
vr->status_reg_addr: 0x20078000
vr->shm_addr: 0x20078004
vr->shm_size: 19744
vr->rx_addr: 0x2007c004
vr->tx_addr: 0x2007c694
```

- On the current `main`
```
#VRING0
vr->vring_size: 16
vr->status_reg_addr: 0x20070000
vr->shm_addr: 0x20070004
vr->shm_size: 17244                           
vr->rx_addr: 0x20074004
vr->tx_addr: 0x200741b4

#VRING1
vr->vring_size: 64                   
vr->status_reg_addr: 0x20078000
vr->shm_addr: 0x20078004
vr->shm_size: 19740
vr->rx_addr: 0x2007c004
vr->tx_addr: 0x2007c694
```

The only difference between `main` and this PR with the alignment set to 4-bytes is the `vr->shm_size` that should not really be important.

Thalley commented:
@carlocaione I am unaware of which version of open-amp our 3rd party controller implements, but it hasn't been updated in nearly a year I believe

Thalley commented:
> > @carlocaione I am unaware of which version of open-amp our 3rd party controller implements, but it hasn't been updated in nearly a year I believe
> 
> I see but my point is: if that 3-rd party controller works with `main` on appcore (it does right?) I don't see why it should not work with this PR + `CONFIG_IPC_SERVICE_STATIC_VRINGS_MEM_ALIGNMENT = 4` on appcore as well, given the experiment in my previous comment.

That is good point. Just in case that my configuration was wrong, you can try to apply https://github.com/nrfconnect/sdk-nrf/blob/main/lib/bin/bt_ll_acs_nrf53/bin/ble5-ctr-rpmsg_3357.hex to the nRF5340 (DK or ADK) and test yourself

iuliana-prodan commented on subsys/ipc/ipc_service/backends/ipc_rpmsg_static_vrings.c:
```diff
@@ -758,6 +758,15 @@ static int backend_init(const struct device *instance)
+	__ASSERT((VDEV_STATUS_SIZE % sys_cache_data_line_size_get()) == 0U,
+		  "VDEV status area but be aligned to the cache line");
+	__ASSERT((VRING_ALIGNMENT % sys_cache_data_line_size_get()) == 0U,
+		  "Static VRINGs but be aligned to the cache line");
```
s/but/must (same above)

arnopo commented on subsys/ipc/ipc_service/lib/Kconfig:
```diff
@@ -11,12 +11,13 @@ config IPC_SERVICE_STATIC_VRINGS
 	int "VRINGs alignment"
 	depends on IPC_SERVICE_STATIC_VRINGS
-	default 4
+	default 32
```
What about setting default to 4 for compatibility with legacy?

hubertmis replied:
What if one peer uses cache and the other one does not? The defaults would differ and the communication could not be established out of the box

carlocaione replied:
aaaalright, switching to `4`.
